apps/publisher/src/hooks/useMediaDevices.ts

```diff
@@ -0,0 +1,76 @@
+import { useEffect, useState } from "react";
+
+type MediaDevices = {
+    cameraList: InputDeviceInfo[];
+    microphoneList: InputDeviceInfo[];
+    setCamera: (device: InputDeviceInfo) => void;
+    setMicrophone: (device: InputDeviceInfo) => void;
+    camera?: InputDeviceInfo;
+    microphone?: InputDeviceInfo;
+    mediaStream?: MediaStream;
+}
+
+const useMediaDevices: () => MediaDevices = () => {
+    const [cameraList, setCameraList] = useState<InputDeviceInfo[]>([]);
+    const [microphoneList, setMicrophoneList] = useState<InputDeviceInfo[]>([]);
+    
+    const [camera, setCamera] = useState<InputDeviceInfo>();
+    const [microphone, setMicrophone] = useState<InputDeviceInfo>();
+
+    const [mediaStream, setMediaStream] = useState<MediaStream | undefined>();
+
+    useEffect(() => {
+        getMediaDevicesList();
```

Comment:
We have to make sure we invoke `getUserMedia` before `getMediaDevicesList`, because `getUserMedia` will ask permission. This is important in Safari. Please test it.